File: src/main/java/com/sothawo/akkabatch/Reader.java
/**
 * $Id: x $
 *
 * Copyright (c) 2014 Jaroso GmbH
 *
 * http://www.jaroso.de
 *
 */
package com.sothawo.akkabatch;

import akka.actor.ActorRef;
import akka.actor.Cancellable;
import com.sothawo.akkabatch.messages.*;
import scala.concurrent.duration.Duration;
import scala.concurrent.duration.FiniteDuration;

import java.io.BufferedReader;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStreamReader;
import java.text.MessageFormat;
import java.util.HashMap;
import java.util.Map;
import java.util.TreeMap;
import java.util.concurrent.TimeUnit;

/**
 * Reader-actor.
 *
 * @author P.J.Meisch (pj.meisch@jaroso.de)
 */
public class Reader extends AkkaBatchActor {
// ------------------------------ FIELDS ------------------------------

    protected static final String CLEANUP_WORKERS = "cleanupWorkers";
    /** the Inbox of the system, for the last message at the end */
    private ActorRef inbox;

    /** to read the input data */
    private BufferedReader reader;

    /** instance of the WorkAvailable message */
    private WorkAvailable workAvailable;

    /** mappign of registered workers to the timestamp of the registration */
    private final Map<ActorRef, Long> workers = new HashMap<>();

    /**
     * actaul work to be done; put in a map, so when the work is rescheduled because of a timeout, it's not duplicated
     */
    private final TreeMap<Long, DoWork> workToBeDone = new TreeMap<>();

    /** maximum number of records allowed in the system */
    private long maxNumRecordsInSystem;

    /** number of records in the system which cause a refill of the internal buffer */
    private long fillLevel;

    /** actal number of records in the system */
    private long actNumRecordsInSystem;

    /** recordId of the next record to read */
    private long recordSerialNo;

    /** number of records read fromm the input file */
    private long numRecordsInInput;

    /** number of records written to the output file */
    private long numRecordsInOutput;

    /** number of processed records, they must not be written out yet */
    private long numRecordsProcessed;

    /** average processing time of a record */
    private long averageProcessingTimeMs;

    /** map from recordIDs to DoWorkInfos */
    private Map<Long, DoWorkInfo> doWorkInfos = new HashMap<>();

    /** mesage for resend */
    private SendAgain resend;

    /** Schedule to cleanup workers */
    private Cancellable cleanupWorkersSchedule;

// ------------------------ CANONICAL METHODS ------------------------

    @Override
    public void onReceive(Object message) throws Exception {
        if (message instanceof Register) {
            registerWorker();
        } else if (message instanceof InitReader) {
            initReader((InitReader) message);
        } else if (message instanceof GetWork) {
            sendWork();
        } else if (message instanceof SendAgain) {
            resendMessages();
        } else if (message instanceof RecordReceived) {
            recordReceived((RecordReceived) message);
        } else if (message instanceof RecordsWritten) {
            recordsWritten((RecordsWritten) message);
        } else if (message.equals(CLEANUP_WORKERS)) {
            cleanupWorkers();
        } else {
            unhandled(message);
        }
    }

    /**
     * removes works which have not reregistered in time from the workers mapder workers Map
     */
    private void cleanupWorkers() {
        Map<ActorRef, Long> cleanWorkers = new HashMap<>();
        // 5x registerInterval means timeout
        long staleTimestamp = System.currentTimeMillis() - (5000 * configApp.getInt("times.registerIntervall"));
        for (Map.Entry<ActorRef, Long> entry : workers.entrySet()) {
            if (entry.getValue() > staleTimestamp) {
                cleanWorkers.put(entry.getKey(), entry.getValue());
            } else {
                log.debug("remove worker " + entry.getKey().path());
            }
        }
        workers.clear();
        workers.putAll(cleanWorkers);
    }

    /**
     * MessageHandler, initializes the Reader and starts processing
     *
     * @param message
     *         init message
     */
    private void initReader(InitReader message) {
        inbox = sender();
        workAvailable = new WorkAvailable();
        maxNumRecordsInSystem = configApp.getLong("numRecords.inSystem");
         fillLevel = (maxNumRecordsInSystem * 9) / 10;
        workToBeDone.clear();
        actNumRecordsInSystem = 0;
        recordSerialNo = 1;
        numRecordsInInput = 0;
        numRecordsInOutput = 0;
        numRecordsProcessed = 0;
        averageProcessingTimeMs = 0;
        Boolean result = true;
        try {
            reader = new BufferedReader(
                    new InputStreamReader(new FileInputStream(message.getInputFilename()), message.getEncoding()));
            fillWorkToBeDone();
        } catch (IOException e) {
            log.error(e, "Initializing Reader");
            result = false;
        }
        log.info(MessageFormat.format("file: {0}, encoding: {1}, Init-result: {2}", message.getInputFilename(),
                                      message.getEncoding(), result));
        if (!result) {
            sender().tell(new WorkDone(false), getSelf());
        } else {
            // resend Message schedulen
            resend = new SendAgain();
            // first resend after 500ms, after that the resend period is calculated dynamically
            getContext().system().scheduler()
                        .scheduleOnce(Duration.create(500, TimeUnit.MILLISECONDS), getSelf(), resend,
                                      getContext().dispatcher(),
                                      null);
        }
    }

    /**
     * removes the corrsepondign record from the map of potentially resendable records and updates the average processing time
     *
     * @param message
     *         contains the id of the processed record
     */
    private void recordReceived(RecordReceived message) {
        Long recordId = message.getId();
        DoWorkInfo doWorkInfo = doWorkInfos.get(recordId);
        if (null != doWorkInfo) {
            doWorkInfos.remove(recordId);
            workToBeDone.remove(recordId);

            long duration = System.currentTimeMillis() - doWorkInfo.getTimestamp();
            if (0 == averageProcessingTimeMs) {
                averageProcessingTimeMs = duration;
            } else if (duration != averageProcessingTimeMs) {
                averageProcessingTimeMs =
                        ((averageProcessingTimeMs * numRecordsProcessed) + duration) / (numRecordsProcessed + 1);
            }
            if(1 > averageProcessingTimeMs) {
                averageProcessingTimeMs = 1;
            }
            numRecordsProcessed++;
        }
    }

    /**
     * MessageHandler, loads new data
     *
     * @param message
     *         contains the number of written records
     * @throws IOException
     */
    private void recordsWritten(RecordsWritten message) throws IOException {
        Long numRecordsWritten = message.getNumRecords();
        actNumRecordsInSystem -= numRecordsWritten;
        fillWorkToBeDone();
        numRecordsInOutput += numRecordsWritten;
        if (null == reader && numRecordsInInput == numRecordsInOutput) {
            // all done
            log.debug(MessageFormat.format("average processing time: {0} ms", averageProcessingTimeMs));
            inbox.tell(new WorkDone(Boolean.TRUE), getSelf());
        }
    }

    /**
     * fills the buffer with records to process and notifies the workers
     */
    private void fillWorkToBeDone() throws IOException {
        // only fill when there are less than the configured records in the system
        if(actNumRecordsInSystem >= fillLevel) {
            return;
        }
        boolean breakout = false;
        while (null != reader && actNumRecordsInSystem < maxNumRecordsInSystem && !breakout) {
            String line = reader.readLine();
            if (line != null) {
                long recordId = recordSerialNo++;
                workToBeDone.put(recordId, new DoWork(recordId, line));
                actNumRecordsInSystem++;
                numRecordsInInput++;
            } else {
                // no more data
                reader.close();
                reader = null;
            }
            if (500 == numRecordsInInput) {
                // break after the first 500 records to get the system started fast
                breakout = true;
            }
        }
        notifyWorkers();
    }

    /**
     * registers the sender as worker and notifies if there is work
     */
    private void registerWorker() {
        ActorRef worker = sender();
        if (!workers.containsKey(worker)) {
            log.info("registered  " + worker.path());
        }
        workers.put(worker, System.currentTimeMillis());
        if (0 < workToBeDone.size()) {
            worker.tell(workAvailable, getSelf());
        }
    }

    /**
     * MessageHandler, resends the timed out records again
     */
    private void resendMessages() {
        // move data with timeout (twice the average processing time)
        long now = System.currentTimeMillis();
        long overdueTimestamp = now - (2 * averageProcessingTimeMs);
        for (Map.Entry<Long, DoWorkInfo> entry : doWorkInfos.entrySet()) {
            DoWorkInfo doWorkInfo = entry.getValue();
            if (doWorkInfo.getTimestamp() <= overdueTimestamp) {
                doWorkInfo.markResend(now);
                workToBeDone.put(entry.getKey(), doWorkInfo.getDoWork());
            }
        }
        notifyWorkers();

        // check for resend in thrice the average processing time
        FiniteDuration interval = Duration.create((0 == averageProcessingTimeMs) ? 500 : (3 * averageProcessingTimeMs),
                                                  TimeUnit.MILLISECONDS);
        getContext().system().scheduler().scheduleOnce(interval, getSelf(), resend, getContext().dispatcher(),
                                                       null);
    }

    /**
     * notifies the workers if there is work to be done
     */
    private void notifyWorkers() {
        if (0 < workToBeDone.size()) {
            for (ActorRef worker : workers.keySet()) {
                worker.tell(workAvailable, getSelf());
            }
        }
    }

    /**
     * sends the next work to the sender
     */
    private void sendWork() {
        if (0 < workToBeDone.size()) {
            DoWork doWork = workToBeDone.firstEntry().getValue();
            Long recordId = doWork.getRecordId();
            workToBeDone.remove(recordId);

            if (!doWorkInfos.containsKey(recordId)) {
                doWorkInfos.put(recordId, new DoWorkInfo(doWork));
            }
            sender().tell(doWork, getSelf());
        }
    }

    @Override
    public void postStop() throws Exception {
        if (null != cleanupWorkersSchedule) {
            cleanupWorkersSchedule.cancel();
            cleanupWorkersSchedule = null;
        }
        super.postStop();
    }

    @Override
    public void preStart() throws Exception {
        super.preStart();
        // regular cleanup message
        cleanupWorkersSchedule = getContext().system().scheduler().schedule(Duration.create(0, TimeUnit.SECONDS),
                                                                            Duration.create(20 * configApp
                                                                                                    .getInt("times.registerIntervall"),
                                                                                            TimeUnit.SECONDS
                                                                            ), getSelf(),
                                                                            CLEANUP_WORKERS,
                                                                            getContext().dispatcher(), getSelf()
        );
    }
}

File: src/main/scala/com/sothawo/akkabatch/scala/Reader.scala
package com.sothawo.akkabatch.scala

import java.io.IOException

import scala.collection.{mutable, Iterator}
import scala.concurrent.duration._
import scala.io.Source

import akka.actor.{ActorRef, Cancellable, Props}
import com.sothawo.akkabatch.scala.messages._

/**
 * Reader-actor.

 * @author P.J. Meisch (pj.meisch@sothawo.com).
 */
class Reader extends AkkaBatchActor {
  protected val CLEANUP_WORKERS: String = "cleanupWorkers"

  /** implicit execution context for the scheduler */
  implicit val ec = context.dispatcher

  /** the Inbox of the system, for the last message at the end */
  private var inbox: ActorRef = _

  /** to read the input data */
  /** need to store the source to be able to close it */
  private var inputSource: Source = _
  private var inputIterator: Iterator[String] = _
  private var inputClosed = true

  /** mapping of registered workers to the timestamp of the registration */
  private var workers = new mutable.HashMap[ActorRef, Long]

  /** actual work to be done; put in a set, so when the work is rescheduled because of a timeout,
    * it's not duplicated. */
  private val workToBeDone = mutable.SortedSet[DoWork]()(DoWorkRecordIdOrdering)

  /** maximum number of records allowed in the system */
  private var maxNumRecordsInSystem: Long = _

  /** number of records in the system which cause a refill of the internal buffer */
  private var fillLevel: Long = _

  /** actual number of records in the system */
  private var actNumRecordsInSystem: Long = _

  /** recordId of the next record to read */
  private var recordSerialNo: Long = _

  /** number of records read fromm the input file */
  private var numRecordsFromInput: Long = _

  /** number of records written to the output file */
  private var numRecordsInOutput: Long = _

  /** number of processed records, they must not be written out yet */
  private var numRecordsProcessed: Long = _

  /** average processing time of a record */
  private var averageProcessingTimeMs: Long = 0L

  /** map from recordIDs to DoWorkInfos */
  private val workInProcess = new mutable.HashMap[Long, DoWorkInfo]

  /** message for resend */
  private val resend: SendAgain = SendAgain()

  /** Schedule to cleanup workers */
  private var cleanupWorkersSchedule: Cancellable = _

  /** instance of the WorkAvailable message */
  private val workAvailable: WorkAvailable = WorkAvailable()

  override def preStart() {
    super.preStart()
    cleanupWorkersSchedule = context.system.scheduler
      .schedule(0 seconds, (20 * appConfig.getInt("times.registerIntervall")) seconds, self, CLEANUP_WORKERS)
  }

  override def postStop() {
    if (null != cleanupWorkersSchedule) {
      cleanupWorkersSchedule.cancel()
      cleanupWorkersSchedule = null
    }
    super.postStop()
  }

  def receive = {
    case _: Register => registerWorker()
    case msg: InitReader => initReader(msg)
    case _: GetWork => sendWork()
    case _: SendAgain => resendMessages()
    case msg: RecordReceived => recordReceived(msg)
    case msg: RecordsWritten => recordsWritten(msg)
    case CLEANUP_WORKERS => cleanupWorkers()
  }

  def registerWorker() {
    if (!(workers contains sender)) log debug (s"registered  ${sender path}")
    workers(sender) = System.currentTimeMillis
    if (workToBeDone.size > 0) sender ! workAvailable
  }

  def initReader(msg: InitReader) {
    inbox = sender
    maxNumRecordsInSystem = appConfig.getLong("numRecords.inSystem")
    fillLevel = (maxNumRecordsInSystem * 9) / 10
    workToBeDone.clear()
    actNumRecordsInSystem = 0
    recordSerialNo = 1
    numRecordsFromInput = 0
    numRecordsInOutput = 0
    numRecordsProcessed = 0
    averageProcessingTimeMs = 0

    var result = true
    try {
      inputSource = Source.fromFile(msg inputFilename, msg encoding)
      inputIterator = inputSource getLines()
      inputClosed = false
      fillWorkToBeDone()
    }
    catch {
      case e: IOException => {
        log.error(e, "Initializing Reader")
        result = false
      }
    }

    log.info(s"file: ${msg.inputFilename}, encoding: ${msg.encoding}, Init-result: ${result}")

    if (!result) {
      sender ! WorkDone(false)
    }
    else {
      context.system.scheduler.scheduleOnce(500 millis, self, resend)
    }
  }

  def sendWork() {
    if (workToBeDone.size > 0) {
      val doWork = workToBeDone head
      val recordId = doWork.recordId
      workToBeDone -= doWork
      // only insert when not contained otherwise the timestamp is modified
      if (!workInProcess.contains(recordId)) workInProcess(recordId) = DoWorkInfo(doWork)
      sender ! doWork
    }
  }

  def resendMessages() {
    // move data with timeout (twice the average processing time)
    val now = System.currentTimeMillis
    val overdueTimestamp = now - (2 * averageProcessingTimeMs)

    for ((recordId, doWorkInfo) <- workInProcess) {
      if (doWorkInfo.timestamp <= overdueTimestamp) {
        doWorkInfo.markResend(now)
        workToBeDone += doWorkInfo.doWork
      }
    }

    // funktional:
    //    workInProcess foreach {
    //      case (recordId, doWorkInfo) => {
    //        if (doWorkInfo.timestamp <= overdueTimestamp) {
    //          doWorkInfo.markResend(now)
    //          workToBeDone.put(recordId, doWorkInfo.doWork)
    //        }
    //      }
    //    }
    notifyWorkers()

    // check for resend in thrice the average processing time
    val interval = if (0 == averageProcessingTimeMs) 500 else averageProcessingTimeMs * 3
    context.system.scheduler.scheduleOnce(interval millis, self, resend)
  }

  def recordReceived(msg: RecordReceived) {
    val recordId = msg.recordId
    if (workInProcess.contains(recordId)) {
      val doWorkInfo = workInProcess(recordId)
      workInProcess -= recordId
      workToBeDone -= doWorkInfo.doWork
      val duration = System.currentTimeMillis - doWorkInfo.timestamp
      if (0 == averageProcessingTimeMs) {
        averageProcessingTimeMs = duration abs
      }
      else if (duration != averageProcessingTimeMs) {
        averageProcessingTimeMs = (((averageProcessingTimeMs * numRecordsProcessed) + duration) / (numRecordsProcessed
          + 1)) max 1
      }
      numRecordsProcessed += 1
    }
  }

  def recordsWritten(msg: RecordsWritten) {
    val numRecordsWritten = msg.numRecords
    actNumRecordsInSystem -= numRecordsWritten
    fillWorkToBeDone()
    numRecordsInOutput += numRecordsWritten
    if (inputClosed && numRecordsFromInput == numRecordsInOutput) {
      log info (s"average processing time: ${averageProcessingTimeMs} ms")
      inbox ! WorkDone(true)
    }
  }

  def cleanupWorkers() {
    // 5x registerInterval means timeout
    val staleTimestamp = System.currentTimeMillis - (5000 * appConfig.getInt("times.registerIntervall"))
    //    workers = workers retain ((actor, timestamp) => if (timestamp > staleTimestamp) true else { log.debug
    // ("remove " +
    //      "worker " + actor.path); false})
    workers = workers retain ((_, timestamp) => timestamp > staleTimestamp)
  }

  def fillWorkToBeDone() {
    // only fill when there are less than the configured records in the system
    if (actNumRecordsInSystem < fillLevel) {
      while (!inputClosed && actNumRecordsInSystem < maxNumRecordsInSystem) {
        if (inputIterator hasNext) {
          val recordId = recordSerialNo;
          workToBeDone += DoWork(recordId, inputIterator.next())
          recordSerialNo += 1
          actNumRecordsInSystem += 1
          numRecordsFromInput += 1
        } else {
          inputSource.close()
          inputClosed = true
        }
      }
      notifyWorkers()
    }
  }

  /**
   * notifies the workers if there is work to be done
   */
  private def notifyWorkers() {
    if (workToBeDone.size > 0) workers.foreach(_._1 ! workAvailable)
  }
}

object Reader {
  def props() = Props(new Reader())
}
